Extend/Controller/OrderController.php

```diff
@@ -0,0 +1,281 @@
+        $oSession = Registry::getSession();
+        $sWdSessionToken = $oSession->getVariable('wdtoken');
+
+        if ($this->_isPaymentFinished($sWdSessionToken, $sWdPaymentRedirect)) {
+            $sShopBaseUrl = $oConfig->getShopUrl();
+            $sLanguageCode = Registry::getLang()->getBaseLanguage();
+
+            $aParams = array(
+                'lang' => $sLanguageCode,
+                'force_sid' => $oSession->getId(),
+                'stoken' => $oSession->getSessionChallengeToken(),
+                'actcontrol' => 'order',
+                'cl' => 'order',
+                'fnc' => 'execute',
+                'sDeliveryAddressMD5' => $this->getDeliveryAddressMD5(),
+                'challenge' => '',
+                'ord_agb' => '1',
+                'oxdownloadableproductsagreement' => '0',
+                'oxserviceproductsagreement' => '0',
+                'wdtoken' => $sWdSessionToken
+            );
+
+            $sParamStr = http_build_query($aParams);
+
+            $sNewUrl = $sShopBaseUrl . 'index.php?' . $sParamStr;
+            header('Location: '.$sNewUrl);
+        }
+    }
+
+    /**
+     * Redirects to the PaymentSDK if the payment is a Module's method
+     *
+     * @inheritdoc
+     *
+     * @return string
+     *
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+     */
+    public function execute()
+    {
+        $oBasket = $this->getSession()->getBasket();
+        $oPayment = oxNew(Payment::class);
+        $oPayment->load($oBasket->getPaymentId());
+
+        if (!$oPayment->isCustomPaymentMethod()) {
+            return parent::execute();
+        }
+
+        $oOrder = oxNew(Order::class);
+        $sOrderId = Registry::getSession()->getVariable('sess_challenge');
+        $bIsOrderLoaded = $oOrder->load($sOrderId);
+        if ($bIsOrderLoaded) {
+            $iSuccess = $oOrder->oxorder__wdoxidee_finalizeorderstate->value;
+        }
+
+        $oConfig = Registry::getConfig();
+        $sWdPaymentRedirect = $oConfig->getRequestParameter('wdtoken');
+        $sWdSessionToken = $this->getSession()->getVariable('wdtoken');
+        if ($this->_isPaymentFinished($sWdSessionToken, $sWdPaymentRedirect)) {
+            return $this->_getNextStep($iSuccess);
+        }
+
+        if ($bIsOrderLoaded) {
+            $oOrder->delete($sOrderId);
+        }
+
+        $oUser = $this->getUser();
+
+        if (!$this->_shouldCreateOrder($oUser)) {
+            return 'payment?payerror=2';
+        }
+
+        if ($oBasket->getProductsCount()) {
+            $oOrder = $this->_createOrder($oBasket, $oUser);
+
+            if (!$oOrder) {
+                $iSuccess = $oOrder->oxorder__wdoxidee_finalizeorderstate->value;
+                return $this->_getNextStep($iSuccess);
+            }
+
+            $this->_handleTransaction($oBasket, $oOrder);
+        }
+
+        //redirect happens before the return
+        return '';
```

Comment:
do we need to die then if the redirect happens before?

Reply:
oxid already handles that